kernel/src/net/socket/util/message_header.rs

```diff
@@ -1,32 +1,190 @@
+        let mut len = 0;
+
+        for msg in msgs.iter() {
+            let header = match msg.write_to(writer) {
+                Ok(header) => header,
+                // This occurs when the buffer is too short or when some page faults cannot be
+                // handled. However, at this point, there is no good way to report the errors to
+                // user space. According to the Linux implementation, it seems okay to silently
+                // ignore errors here.
+                Err(_) => {
+                    warn!("setting MSG_CTRUNC is not supported");
+                    break;
+                }
+            };
+
+            len += header.total_len();
+
+            let padding_len = header.padding_len().min(writer.avail());
+            writer.skip(padding_len);
+            len += padding_len;
+        }
+
+        len
+    }
+
+    fn write_to(&self, writer: &mut VmWriter) -> Result<CControlHeader> {
+        match self {
+            Self::Unix(msg) => msg.write_to(writer),
+        }
+    }
+}
+
+/// `cmsghdr` in Linux.
+///
+/// Reference: <https://elixir.bootlin.com/linux/v6.13/source/include/uapi/linux/netlink.h#L52>.
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Pod)]
+pub struct CControlHeader {
+    /// Data byte count, including hdr
+    len: usize,
+    /// Originating protocol
+    level: i32,
+    /// Protocol-specific type
+    type_: i32,
+}
+
+/// Alignment of control messages.
+///
+/// Reference: <https://elixir.bootlin.com/linux/v6.13/source/include/linux/socket.h#L119>.
+const CMSG_ALIGN: usize = size_of::<usize>();
+
+impl CControlHeader {
+    /// Creates a control message header with the level, type, and payload length.
+    pub fn from_payload_len(level: CSocketOptionLevel, type_: i32, payload_len: usize) -> Self {
+        Self {
+            len: payload_len + size_of::<Self>(),
+            level: level as i32,
+            type_,
+        }
+    }
+
+    /// Computes the payload length from the total length.
+    pub fn payload_len_from_total(total_len: usize) -> Result<usize> {
+        total_len.checked_sub(size_of::<Self>()).ok_or_else(|| {
+            Error::with_message(Errno::EINVAL, "the control message buffer is too small")
+        })
+    }
+
+    /// Returns the level of the control message.
+    pub fn level(&self) -> Option<CSocketOptionLevel> {
+        CSocketOptionLevel::try_from(self.level).ok()
+    }
+
+    /// Returns the type of the control message.
+    pub fn type_(&self) -> i32 {
+        self.type_
+    }
+
+    /// Returns the payload length of the control message.
+    pub fn payload_len(&self) -> usize {
+        self.len - size_of::<Self>()
+    }
+
+    /// Returns the length of the control message (payload + header, excluding paddings).
+    pub fn total_len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns the length of the control message (payload + header, including paddings).
+    pub fn total_len_with_padding(&self) -> usize {
```

Comment:
Can this method be made private? It doesn’t appear to be used outside the module.